Refactor some enums in EditorState to enum classes

jesxilin opened the pull request:
#### 72fad77b89f09c7567e56e5da0923fc3d4534a2b
<pre>
Refactor some enums in EditorState to enum classes
<a href="https://bugs.webkit.org/show_bug.cgi?id=256930">https://bugs.webkit.org/show_bug.cgi?id=256930</a>
rdar://109487509

Reviewed by Chris Dumez.

Refactored EditorState enums ListType, TextAlignment, and TypingAttributes to enum classes. Changed TypingAttributes to be TypingAttribute.
Changed typingAttributes in PostLayoutData to be an OptionSet.

* Source/WebKit/Shared/EditorState.cpp:
(WebKit::operator&lt;&lt;):
* Source/WebKit/Shared/EditorState.h:
* Source/WebKit/Shared/EditorState.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(dictionaryRepresentationForEditorState):
(nsTextAlignment):
* Source/WebKit/UIProcess/API/glib/WebKitEditorState.cpp:
(webkitEditorStateChanged):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView textStylingAtPosition:inDirection:]):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(-[WKTextListTouchBarViewController _selectList:]):
(-[WKTextListTouchBarViewController setCurrentListType:]):
(WebKit::nsTextAlignmentFromTextAlignment):
(WebKit::WebViewImpl::updateTextTouchBar):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::getPlatformEditorStateCommon const):

Canonical link: <a href="https://commits.webkit.org/264439@main">https://commits.webkit.org/264439@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/def5477e80787d5b44d461c86dd70c6783b344ed

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7598 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7866 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8050 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9241 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7787 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/9825 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7793 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10660 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7731 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/8845 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7009 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9350 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6154 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14620 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/7327 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7043 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10443 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7536 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6154 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6874 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1815 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11085 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7267 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/Shared/EditorState.h:
```diff
@@ -48,23 +48,23 @@ class TextStream;
 };
 
-enum TextAlignment {
+enum class TextAlignment : uint16_t {
```
```suggestion
enum class TextAlignment : uint8_t {
```

whsieh commented on Source/WebKit/Shared/EditorState.h:
```diff
@@ -48,23 +48,23 @@ class TextStream;
 namespace WebKit {
 
-enum TypingAttributes {
+enum class TypingAttribute : uint16_t {
```
```suggestion
enum class TypingAttribute : uint8_t {
```

whsieh replied:
Nit - I think we can probably rename `AttributeBold` to just `Bold` as well:

```
enum class TypingAttributes : uint8_t {
    None = 0,
    Bold = 1,
    Italics = 2,
    Underline = 4,
    StrikeThrough = 8
};
```

whsieh commented on Source/WebKit/Shared/EditorState.h:
```diff
@@ -48,23 +48,23 @@ class TextStream;
 };
 
-enum ListType {
+enum class ListType : uint16_t {
```
```suggestion
enum class ListType : uint8_t {
```

rr-codes commented on Source/WebKit/Shared/EditorState.cpp:
```diff
@@ -62,17 +62,17 @@ TextStream& operator<<(TextStream& ts, const EditorState& editorState)
-        if (editorState.postLayoutData->textAlignment != NoAlignment)
-            ts.dumpProperty("textAlignment", editorState.postLayoutData->textAlignment);
+        if (editorState.postLayoutData->textAlignment != TextAlignment::NoAlignment)
+            ts.dumpProperty("textAlignment", static_cast<unsigned>(editorState.postLayoutData->textAlignment)); // giving big error
```
this should use `WTF::enumToUnderlyingType` (also remove the comment 😛 )

rr-codes commented on Source/WebKit/Shared/EditorState.h:
```diff
@@ -48,23 +48,23 @@ class TextStream;
 
-enum TypingAttributes {
+enum class TypingAttribute : uint16_t {
     AttributeNone = 0,
```
these enum cases should probably have their Attribute prefix removed to reduce redundancy

whsieh reviewed (commented):
Great start! Just a few minor comments.

rr-codes commented on Source/WebKit/Shared/EditorState.h:
```diff
@@ -48,23 +48,23 @@ class TextStream;
 
-enum ListType {
+enum class ListType : uint16_t {
     NoList = 0,
```
this should probably be `None`

cdumez commented on Source/WebKit/Shared/EditorState.h:
```diff
@@ -48,24 +48,23 @@ class TextStream;
-    CenterAlignment = 3,
-    JustifiedAlignment = 4,
+enum class TextAlignment : uint8_t {
+    NoAlignment,
```
May be nicer to call this `None` now that we have to specify the enum name (and the name already says "Alignment").

hortont424 replied:
What even does TextAlignment::None mean? How can you have text with no alignment!

rr-codes replied:
it means `NSTextAlignmentNatural` apparently!

rr-codes replied:
so maybe rename from NoAlignment to Natural?

cdumez commented on Source/WebKit/Shared/EditorState.h:
```diff
@@ -88,12 +87,12 @@ struct EditorState {
 
     struct PostLayoutData {
-        uint32_t typingAttributes { AttributeNone };
+        OptionSet<TypingAttribute> typingAttributes { };
```
Does it build without the `{ }`? We should omit if possible.

cdumez commented on Source/WebKit/Shared/EditorState.cpp:
```diff
@@ -62,17 +62,17 @@ TextStream& operator<<(TextStream& ts, const EditorState& editorState)
-        if (editorState.postLayoutData->textAlignment != NoAlignment)
-            ts.dumpProperty("textAlignment", editorState.postLayoutData->textAlignment);
+        if (editorState.postLayoutData->textAlignment != TextAlignment::NoAlignment)
+            ts.dumpProperty("textAlignment", WTF::enumToUnderlyingType(editorState.postLayoutData->textAlignment));
```
`WTF::` can be omitted.

cdumez commented on Source/WebKit/Shared/EditorState.cpp:
```diff
@@ -62,17 +62,17 @@ TextStream& operator<<(TextStream& ts, const EditorState& editorState)
-        if (editorState.postLayoutData->enclosingListType != NoList)
-            ts.dumpProperty("enclosingListType", editorState.postLayoutData->enclosingListType);
+        if (editorState.postLayoutData->enclosingListType != ListType::None)
+            ts.dumpProperty("enclosingListType", WTF::enumToUnderlyingType(editorState.postLayoutData->enclosingListType));
```
`WTF::` can be omitted.

cdumez reviewed (approved):
LGTM with nit fixes.

pxlcoder commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -5800,9 +5800,9 @@ static NSTextAlignment nsTextAlignmentFromTextAlignment(TextAlignment textAlignm
-            [m_textTouchBarItemController setTextIsUnderlined:(bool)(m_page->editorState().postLayoutData->typingAttributes & AttributeUnderline)];
+            [m_textTouchBarItemController setTextIsBold:(bool)(m_page->editorState().postLayoutData->typingAttributes & TypingAttribute::Bold)];
+            [m_textTouchBarItemController setTextIsItalic:(bool)(m_page->editorState().postLayoutData->typingAttributes &  TypingAttribute::Italics)];
+            [m_textTouchBarItemController setTextIsUnderlined:(bool)(m_page->editorState().postLayoutData->typingAttributes &  TypingAttribute::Underline)];
```
Since this is now an `OptionSet`, we should use `.contains` rather than `&`, and remove the cast to `bool`.

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -1823,9 +1823,9 @@ - (void)createWebArchiveDataWithCompletionHandler:(void (^)(NSData *, NSError *)
-        @"underline": postLayoutData.typingAttributes & WebKit::AttributeUnderline ? @YES : @NO,
+        @"bold": postLayoutData.typingAttributes & WebKit::TypingAttribute::Bold ? @YES : @NO,
+        @"italic": postLayoutData.typingAttributes & WebKit::TypingAttribute::Italics ? @YES : @NO,
+        @"underline": postLayoutData.typingAttributes & WebKit::TypingAttribute::Underline ? @YES : @NO,
```
Another place we should use `.contains` rather than `&`.

pxlcoder commented on Source/WebKit/UIProcess/API/glib/WebKitEditorState.cpp:
```diff
@@ -123,13 +123,13 @@ void webkitEditorStateChanged(WebKitEditorState* editorState, const EditorState&
+    if (postLayoutData.typingAttributes & WebKit::TypingAttribute::Underline)
         typingAttributes |= WEBKIT_EDITOR_TYPING_ATTRIBUTE_UNDERLINE;
-    if (postLayoutData.typingAttributes & AttributeStrikeThrough)
+    if (postLayoutData.typingAttributes & WebKit::TypingAttribute::StrikeThrough)
```
Another place we should use `.contains` rather than `&`.

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -3880,9 +3880,9 @@ - (NSDictionary *)textStylingAtPosition:(UITextPosition *)position inDirection:(
+    if (typingAttributes & WebKit::TypingAttribute::Bold)
         symbolicTraits |= kCTFontBoldTrait;
-    if (typingAttributes & WebKit::AttributeItalics)
+    if (typingAttributes & WebKit::TypingAttribute::Italics)
```
Another place we should use `.contains` rather than `&`.

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -3896,7 +3896,7 @@ - (NSDictionary *)textStylingAtPosition:(UITextPosition *)position inDirection:(
         [result setObject:(id)font.get() forKey:NSFontAttributeName];
     
-    if (typingAttributes & WebKit::AttributeUnderline)
+    if (typingAttributes & WebKit::TypingAttribute::Underline)
```
`contains`

pxlcoder commented on Source/WebKit/WebProcess/WebPage/glib/WebPageGLib.cpp:
```diff
@@ -126,22 +126,22 @@ void WebPage::getPlatformEditorState(LocalFrame& frame, EditorState& result) con
     if (selection.isRange()) {
         if (editor.selectionHasStyle(CSSPropertyFontWeight, "bold"_s) == TriState::True)
-            postLayoutData.typingAttributes |= AttributeBold;
+            postLayoutData.typingAttributes |= TypingAttribute::Bold;
```
`|=` should be `add` as you did above.

rr-codes commented on Source/WebKit/Shared/EditorState.h:
```diff
@@ -48,24 +48,23 @@ class TextStream;
-    CenterAlignment = 3,
-    JustifiedAlignment = 4,
+enum class TextAlignment : uint8_t {
+    Neutral,
```
This should be "Natural" not "Neutral" (to match `NSTextAlignmentNatural`)

Commit:
Refactor some enums in EditorState to enum classes
https://bugs.webkit.org/show_bug.cgi?id=256930
rdar://109487509

Reviewed by Chris Dumez.

Refactored EditorState enums ListType, TextAlignment, and TypingAttributes to enum classes. Changed TypingAttributes to be TypingAttribute.
Changed typingAttributes in PostLayoutData to be an OptionSet.

* Source/WebKit/Shared/EditorState.cpp:
(WebKit::operator<<):
* Source/WebKit/Shared/EditorState.h:
* Source/WebKit/Shared/EditorState.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(dictionaryRepresentationForEditorState):
(nsTextAlignment):
* Source/WebKit/UIProcess/API/glib/WebKitEditorState.cpp:
(webkitEditorStateChanged):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView textStylingAtPosition:inDirection:]):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(-[WKTextListTouchBarViewController _selectList:]):
(-[WKTextListTouchBarViewController setCurrentListType:]):
(WebKit::nsTextAlignmentFromTextAlignment):
(WebKit::WebViewImpl::updateTextTouchBar):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::getPlatformEditorStateCommon const):

Canonical link: https://commits.webkit.org/264439@main
